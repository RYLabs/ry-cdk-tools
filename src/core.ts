export * from "./base_stacks/base_stack";
export { default as BaseStack } from "./base_stacks/base_stack";
export { default as Conventions } from "./utils/conventions";
