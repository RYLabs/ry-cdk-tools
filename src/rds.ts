export * from "./stacks/rds_stack";
