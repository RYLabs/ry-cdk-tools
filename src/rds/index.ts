export { default as RdsStack } from "../stacks/rds_stack";
export * from "../stacks/rds_stack";

export { default as PostgresInstance } from "./postgres_instance";
export * from "./postgres_instance";

export { default as RyDatabaseInstance } from "./ry_database_instance";
export * from "./ry_database_instance";
