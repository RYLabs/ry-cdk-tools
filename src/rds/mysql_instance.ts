import RyDatabaseInstance, {
  BaseRyDatabaseInstanceProps,
} from "./ry_database_instance";
import { Construct } from "@aws-cdk/core";
import { DatabaseInstanceEngine } from "@aws-cdk/aws-rds";

export interface MysqlInstanceProps extends BaseRyDatabaseInstanceProps {}

export default class PostgresInstance extends RyDatabaseInstance {
  constructor(scope: Construct, id: string, props: MysqlInstanceProps) {
    super(scope, id, {
      ...props,
      engine: DatabaseInstanceEngine.MYSQL,
    });
  }
}
