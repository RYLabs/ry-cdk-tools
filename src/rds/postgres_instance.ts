import RyDatabaseInstance, {
  BaseRyDatabaseInstanceProps,
} from "./ry_database_instance";
import { Construct } from "@aws-cdk/core";
import { DatabaseInstanceEngine } from "@aws-cdk/aws-rds";

export interface PostgresInstanceProps extends BaseRyDatabaseInstanceProps {}

export default class PostgresInstance extends RyDatabaseInstance {
  constructor(scope: Construct, id: string, props: PostgresInstanceProps) {
    super(scope, id, {
      ...props,
      engine: DatabaseInstanceEngine.POSTGRES,
    });
  }
}
