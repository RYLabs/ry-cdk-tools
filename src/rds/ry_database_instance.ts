import {
  DatabaseInstance,
  DatabaseInstanceProps,
  DatabaseInstanceEngine,
} from "@aws-cdk/aws-rds";
import { SecretValue, Construct, Duration } from "@aws-cdk/core";
import {
  SecurityGroup,
  InstanceType,
  InstanceClass,
  InstanceSize,
} from "@aws-cdk/aws-ec2";
import { Secret } from "@aws-cdk/aws-secretsmanager";
import Conventions from "../constructs/conventions";

export const instanceDefaults = {
  backupRetention: Duration.days(7),
  instanceClass: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
};

export type AltDatabaseInstanceProps = Omit<
  DatabaseInstanceProps,
  "masterUsername"
>;

export type BaseRyDatabaseInstanceProps = Omit<
  RyDatabaseInstanceProps,
  "engine"
>;

export interface RyDatabaseInstanceProps extends AltDatabaseInstanceProps {
  readonly conventions: Conventions;
  readonly masterUsername?: string;
}

export default class RyDatabaseInstance extends DatabaseInstance {
  masterUsername: string;
  masterPassword: SecretValue;
  databaseName: string;
  securityGroup: SecurityGroup;

  constructor(scope: Construct, id: string, props: RyDatabaseInstanceProps) {
    const {
      conventions,

      masterUsername = `${conventions.eqn("camel")}DbUser`,
      masterUserPassword = new Secret(
        scope,

        // this name appears to get truncated by Secrets Manager, so best to best
        // to keep it short.
        `${conventions.eqn("path")}dbMasterPassword`,

        {
          generateSecretString: {
            includeSpace: false,

            // RDS doesn't like @ / or \ values.  Elasticbeanstalk will choke on
            // backticks, so to be safe we're just going to avoid all quote
            // characters.
            excludeCharacters: "{}[]\"'`@/\\",

            passwordLength: 16,
            secretStringTemplate: JSON.stringify({
              username: `${conventions.eqn("camel")}DbUser`,
            }),
            generateStringKey: "password",
          },
        }
      ).secretValueFromJson("password"),

      databaseName = conventions.eqn("underscore"),
      instanceIdentifier = conventions.eqn(),
      engine,
    } = props;

    super(scope, id, {
      ...instanceDefaults,
      ...props,

      masterUsername,
      masterUserPassword,
      databaseName,
      instanceIdentifier,
    });

    this.masterUsername = masterUsername;
    this.masterPassword = masterUserPassword;
    this.databaseName = databaseName;
  }
}
