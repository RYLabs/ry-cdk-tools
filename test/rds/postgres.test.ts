import { SynthUtils } from "@aws-cdk/assert";
import { Stack } from "@aws-cdk/core";
import { Vpc } from "@aws-cdk/aws-ec2";

import { PostgresInstance } from "../../lib/rds";
import Conventions from "../../lib/constructs/conventions";

test("creates a Postgres instance", () => {
  const stack = new Stack();
  const vpc = new Vpc(stack, "VPC");
  const conventions = new Conventions("app", "dev");
  new PostgresInstance(stack, "Postgres", {
    vpc,
    conventions,
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
