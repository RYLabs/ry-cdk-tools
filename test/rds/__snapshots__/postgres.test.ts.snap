// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a Postgres instance 1`] = `
Object {
  "Resources": Object {
    "Postgres9DC8BB04": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AllocatedStorage": "100",
        "BackupRetentionPeriod": 7,
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.t3.micro",
        "DBInstanceIdentifier": "app-dev",
        "DBName": "app_dev",
        "DBSubnetGroupName": Object {
          "Ref": "PostgresSubnetGroup9F8A4D6E",
        },
        "DeletionProtection": true,
        "Engine": "postgres",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "appdevdbMasterPasswordBF210889",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": "appDevDbUser",
        "StorageType": "gp2",
        "VPCSecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "PostgresSecurityGroup670916BF",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Retain",
    },
    "PostgresSecurityGroup670916BF": Object {
      "Properties": Object {
        "GroupDescription": "Security group for Postgres database",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "PostgresSubnetGroup9F8A4D6E": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnet group for Postgres database",
        "SubnetIds": Array [
          Object {
            "Ref": "VPCPrivateSubnet1Subnet8BCA10E0",
          },
          Object {
            "Ref": "VPCPrivateSubnet2SubnetCFCDAA7A",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "VPCB9E5F0B4": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VPCIGWB7E252D3": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VPCPrivateSubnet1DefaultRouteAE1D6490": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VPCPublicSubnet1NATGatewayE0556630",
        },
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet1RouteTableBE8A6027",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPrivateSubnet1RouteTableAssociation347902D1": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet1RouteTableBE8A6027",
        },
        "SubnetId": Object {
          "Ref": "VPCPrivateSubnet1Subnet8BCA10E0",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPrivateSubnet1RouteTableBE8A6027": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPrivateSubnet1Subnet8BCA10E0": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PrivateSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPrivateSubnet2DefaultRouteF4F5CFD2": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VPCPublicSubnet2NATGateway3C070193",
        },
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet2RouteTable0A19E10E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPrivateSubnet2RouteTable0A19E10E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPrivateSubnet2RouteTableAssociation0C73D413": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet2RouteTable0A19E10E",
        },
        "SubnetId": Object {
          "Ref": "VPCPrivateSubnet2SubnetCFCDAA7A",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPrivateSubnet2SubnetCFCDAA7A": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PrivateSubnet2",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPublicSubnet1DefaultRoute91CEF279": Object {
      "DependsOn": Array [
        "VPCVPCGW99B986DC",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet1RouteTableFEE4B781",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPublicSubnet1EIP6AD938E8": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VPCPublicSubnet1NATGatewayE0556630": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VPCPublicSubnet1EIP6AD938E8",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VPCPublicSubnet1SubnetB4246D30",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VPCPublicSubnet1RouteTableAssociation0B0896DC": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet1RouteTableFEE4B781",
        },
        "SubnetId": Object {
          "Ref": "VPCPublicSubnet1SubnetB4246D30",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPublicSubnet1RouteTableFEE4B781": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPublicSubnet1SubnetB4246D30": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPublicSubnet2DefaultRouteB7481BBA": Object {
      "DependsOn": Array [
        "VPCVPCGW99B986DC",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet2RouteTable6F1A15F1",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPublicSubnet2EIP4947BC00": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VPCPublicSubnet2NATGateway3C070193": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VPCPublicSubnet2EIP4947BC00",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VPCPublicSubnet2Subnet74179F39",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VPCPublicSubnet2RouteTable6F1A15F1": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPublicSubnet2RouteTableAssociation5A808732": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet2RouteTable6F1A15F1",
        },
        "SubnetId": Object {
          "Ref": "VPCPublicSubnet2Subnet74179F39",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPublicSubnet2Subnet74179F39": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC/PublicSubnet2",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCVPCGW99B986DC": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "appdevdbMasterPasswordBF210889": Object {
      "Properties": Object {
        "Description": "password for app-dev database",
        "GenerateSecretString": Object {
          "ExcludeCharacters": "{}[]\\"'\`@/\\\\",
          "GenerateStringKey": "password",
          "IncludeSpace": false,
          "PasswordLength": 16,
          "SecretStringTemplate": "{\\"username\\":\\"appDevDbUser\\"}",
        },
      },
      "Type": "AWS::SecretsManager::Secret",
    },
  },
}
`;
